### PR TITLE
[InstSimplify] Fold converted urem to 0 if there's no overlapping bits

### PR DESCRIPTION
**Pull request by huntergr-arm**
When folding urem instructions we can end up not recognizing that
the output will always be 0 due to Value*s being different, despite
generating the same data (in this case, 2 different calls to vscale).

This patch recognizes the (x << N) & (add (x << M), -1) pattern that
instcombine replaces urem with after the two vscale calls have been
reduced to one via CSE, then replaces with 0 when x is a non-zero
power of 2 and N >= M.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Alive2: https://alive2.llvm.org/ce/z/PdD_x2

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Comment by goldsteinn** [comment]
> Alive2: https://alive2.llvm.org/ce/z/PdD_x2

Works for pow2 or zero: https://alive2.llvm.org/ce/z/UeesRf

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:2671]: or zero can be true: https://alive2.llvm.org/ce/z/UeesRf

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:2669]: You don't need these OneUse checks. This fold won't create more insts.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:2671]: Also can you add `/*Depth*/` comment.

File: llvm/test/Transforms/InstCombine/po2-shift-add-and-to-zero.ll
Comment (huntergr-arm) [reply llvm/test/Transforms/InstCombine/po2-shift-add-and-to-zero.ll:1]: Done in e397a465a87d88f6bc94392a925f296b52345c2f

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (huntergr-arm) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Thanks for checking on that, done.

File: llvm/test/Transforms/InstCombine/po2-shift-add-and-to-zero.ll
Comment (huntergr-arm) [reply llvm/test/Transforms/InstCombine/po2-shift-add-and-to-zero.ll]: Done, though the vscale_range attribute is required (vscale isn't known to be a power of 2 without that being set). I also moved the test to the InstSimplify test directory.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (huntergr-arm) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: I had to rearrange things a bit when moving to simplifyAndInst, since 'I' was no longer available (just Op0/Op1). I was able to use m_Specific as a result.

File: llvm/test/Transforms/InstSimplify/po2-shift-add-and-to-zero.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstSimplify/po2-shift-add-and-to-zero.ll:42]: Please drop unused `nsw/nuw` flags in the following `shl/add` insts.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
Please add a test for `X` is a power of 2 or zero.
```
define i64 @test_pow2_or_zero(i64 %arg) {
  %neg = sub i64 0, %arg
  %x = and i64 %neg, %arg
  %shl1 = shl i64 %x, 4
  %shl2 = shl i64 %x, 3
  %mask = add i64 %shl2, -1
  %rem = and i64 %shl1, %mask
  ret i64 %rem
}
```
Alive2: https://alive2.llvm.org/ce/z/_e3zLi
Wonder if we need some negative tests here.

**Comment by huntergr-arm** [comment]
Precommitted extra tests in 34f83e86b4b70427969439683360a819a13e6f49 (including a negative test where vscale doesn't have a defined range)

**Review by dtcxzyw — APPROVED** [review]
LGTM. Waiting for additional approval from other reviewers.

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:2113]: Is there a negative test where this condition fails?

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:2120]: As far as I can see, there is no test for the commuted version? It looks like all your test have the mask on the RHS.

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:2107]: This pattern is basically the same as the `and 2^x-1, 2^C --> 0 where x <= C` one above -- but unfortunately I don't see any way to combine them with the analysis helpers we currently have available. So I think your approach of just adding this extra pattern is fine.

Comment (huntergr-arm) [reply llvm/lib/Analysis/InstructionSimplify.cpp:2107]: Yes, I realized it was very similar, so put my code just after. An 'm_Maybe' would be nice, but as you say it doesn't exist yet.

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (huntergr-arm) [reply llvm/lib/Analysis/InstructionSimplify.cpp:2120]: Swapped the mask for the second positive test.

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (huntergr-arm) [reply llvm/lib/Analysis/InstructionSimplify.cpp:2113]: I added a negative test in the second precommit to check for failure with a non-power-of-2 input, but I've now added tests with a right shift instead of left, with the larger shift being on the -1 side, and for an add of something other than -1. Are there any others you would like?

**Comment by nikic** [comment]
Could you please rebase over https://github.com/llvm/llvm-project/commit/ebb8ffde94476333cb2e95aacfd2e023d7112efb and move the fold into the simplifyAndCommutative() function, so it does not have to be repeated twice?

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (nikic) [reply llvm/lib/Analysis/InstructionSimplify.cpp:2113]: I think that's all we need, thanks!

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
